### PR TITLE
Fixes several path tests according to the PartiQL Specification

### PR DESCRIPTION
**Pull request by johnedquinn**
# Description
- Fixes several path tests according to the PartiQL Specification

Note: Please click "Hide Whitespace" on the Files Changed tab. The file was formatted oddly, so I allowed my IDE to auto-format.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-tests-data/eval/primitives/path.ion [hunk 1]
```diff
@@ -131,161 +130,230 @@ envs::{
...
-      evalMode:[EvalModeCoerce, EvalModeError],
-      result:EvaluationSuccess,
-      output:true
+    name: "pathNullDotName",
+    statement: "(NULL).a IS MISSING",
+    assert: {
+      evalMode: [
+        EvalModeCoerce,
+        EvalModeError
+      ],
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:194]: This is really weird...

Build up on the previous test cases: 

If MISSING.a is MISSING failed evalMode: EvalModeError, why should NULL.a succeed? 

I understand you can coerce NULL to any type, but it does seem weird to me ....

File: partiql-tests-data/eval/primitives/path.ion [hunk 2]
```diff
@@ -353,412 +433,480 @@ path::[
...
+      {
+        evalMode: EvalModeCoerce,
+        result: EvaluationSuccess,
+        output: $bag::[
+          "hello"
+        ]
+      },
+      {
+        evalMode: EvalModeError,
+        result: EvaluationFail
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:448]: Could you please elaborate on why this failed? 

```
SELECT VALUE v1
FROM s as v1
```

File: partiql-tests-data/eval/primitives/path.ion [hunk 3]
```diff
@@ -353,412 +433,480 @@ path::[
...
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:465]: Could you please elaborate on why this failed?

```
SELECT VALUE v1 
FROM UNPIVOT s as v1
```

Comment (yliuuuu) [reply partiql-tests-data/eval/primitives/path.ion:465]: Spec section 5.1.1 - Iteration over a scalar value

FROM s AS V
where s is a scalar value. Then s coerces into the bag << s >>. 

This section seems to be mode independent, i.e., coercion always happens.

File: partiql-tests-data/eval/primitives/path.ion [hunk 4]
```diff
@@ -353,412 +433,480 @@ path::[
...
+      {
+        evalMode: EvalModeCoerce,
+        result: EvaluationSuccess,
+        output: $bag::[
+          100
+        ]
+      },
+      {
+        evalMode: EvalModeError,
+        result: EvaluationFail
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:482]: Same as above.

File: partiql-tests-data/eval/primitives/path.ion [hunk 5]
```diff
@@ -353,412 +433,480 @@ path::[
...
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:499]: Same as above.

File: partiql-tests-data/eval/primitives/path.ion [hunk 6]
```diff
@@ -353,412 +433,480 @@ path::[
...
             }
           }
-        }
-      ]
-    }
+        ]
+      },
+      {
+        evalMode: EvalModeError,
+        result: EvaluationFail
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:525]: Same as above.

File: partiql-tests-data/eval/primitives/path.ion [hunk 7]
```diff
@@ -353,412 +433,480 @@ path::[
...
+        evalMode: EvalModeCoerce,
+        result: EvaluationSuccess,
+        output: $bag::[
+          5,
+          6
+        ]
+      },
+      {
+        evalMode: EvalModeError,
+        result: EvaluationFail
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:543]: Same as above.

File: partiql-tests-data/eval/primitives/path.ion [hunk 8]
```diff
@@ -353,412 +433,480 @@ path::[
...
-      ],
-      output:"this is unambiguous_lookup_field"
-    }
+    name: "subscript with variable in lowercase",
+    statement: "test_struct[unambiguous_variable]",
+    assert: [
+      {
+        result: EvaluationSuccess,
+        evalMode: EvalModeCoerce,
+        output: $missing::null
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:756]: Tuple navigation with array notation The expression a[s] is a shorthand for the tuple path navigation a.s when
the expression s is either (a) a string literal or (b) an expression that is explicitly CAST into a string. For example:
{'a': 1, 'b': 2}['a'] ⇔ {'a': 1, 'b':2}.'a' → 1

Not sure what the single quoted 'a' means in our spec. but curious why you think this should return a MISSING...

File: partiql-tests-data/eval/primitives/path.ion [hunk 9]
```diff
@@ -353,412 +433,480 @@ path::[
...
-      ],
-      output:"this is unambiguous_lookup_field"
-    }
+    name: "subscript with variable in uppercase",
+    statement: "test_struct[UNAMBIGUOUS_VARIABLE]",
+    assert: [
+      {
+        result: EvaluationSuccess,
+        evalMode: EvalModeCoerce,
+        output: $missing::null
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:773]: Same as above.

File: partiql-tests-data/eval/primitives/path.ion [hunk 10]
```diff
@@ -353,412 +433,480 @@ path::[
...
-      output:"this is unambiguous_lookup_field"
-    }
+    name: "subscript with variable in mixed case",
+    statement: "test_struct[uNaMbiGuOuS_VaRiAbLE]",
+    assert: [
+      {
+        result: EvaluationSuccess,
+        evalMode: EvalModeCoerce,
+        output: $missing::null
+      },
```
Comment (yliuuuu) [inline partiql-tests-data/eval/primitives/path.ion:789]: Same as above.

**Review by yliuuuu — COMMENTED** [review]
I tagged duplicated comments "Same as above" to prevent losing track of the test case.

File: partiql-tests-data/eval/primitives/path.ion (hunk 2 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:448]: From PartiQL Specification:

```
The expression e[∗] reduces to (i.e., is equivalent to):
SELECT VALUE v FROM e AS v
```

And,

> In the following cases the expression in the FROM clause item has the wrong type. Under the type checking option, all of these cases raise an error and the query fails. Under the permissive option, the cases proceed as follows
> -- Section 5.1.1: Mistyping Cases

Therefore,
```
<scalar>[*] = SELECT VALUE v FROM <scalar> AS v --> FROM clause item has the wrong type.
```

File: partiql-tests-data/eval/primitives/path.ion (hunk 3 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:465]: At the top of Section 5.1.1:
> In the following cases the expression in the FROM clause item has the wrong type. **Under the type checking option, all of these cases raise an error and the query fails. Under the permissive option, the cases proceed as follows**...

File: partiql-tests-data/eval/primitives/path.ion (hunk 4 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:482]: See my comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 5 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:499]: See my comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 6 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:525]: See my comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 7 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:543]: See my comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 8 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:756]: > when the expression s is either (a) a string literal or (b) an expression that is explicitly CAST into a string... If s is not a string literal or an expression that is cast into a string, then a[s] is evaluated as an array path navigation.

So, `some_struct['a']` (literal string `'a'`) results in a path expression. Similarly, `some_struct[CAST('a' || 'b' AS STRING)]` (explicit cast) results in a path expression.

"If s is not a string literal or an expression that is cast into a string" (like variable references, plus operations, etc), "then a[s] is evaluated as an array path navigation." All of the below evaluate as an array path navigation:
1. `some_struct[some_variable]`
2. `some_struct[1 + 1]`
3. `some_struct['hello' || ', world']`

File: partiql-tests-data/eval/primitives/path.ion (hunk 10 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:789]: See comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 9 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:773]: See comment.

File: partiql-tests-data/eval/primitives/path.ion (hunk 8 above)
Comment (yliuuuu) [reply partiql-tests-data/eval/primitives/path.ion:756]: Gotya. So `test_struct['unambiguous_variable']` => 'this is unambiguous_lookup_field'`. And the test case is missing single quote here. Is this understanding correct?

File: partiql-tests-data/eval/primitives/path.ion (hunk 1 above)
Comment (yliuuuu) [reply partiql-tests-data/eval/primitives/path.ion:194]: Nonblocking: but we can think of path navigation as a function that is not nullCall?

**Review by yliuuuu — APPROVED** [review]
Thanks for clarifying.

File: partiql-tests-data/eval/primitives/path.ion (hunk 8 above)
Comment (johnedquinn) [reply partiql-tests-data/eval/primitives/path.ion:756]: Well, I think the intention of the test is a bit different. `unambiguous_variable` is a global variable pointing to some string, and I believe the original intention was to test that the underlying value that the variable points to can be used to key into structs -- though that assumption is wrong according to the specification.

**Pull request merged**